In [9]:
def ReadFileIntoList(filename):
    with open(filename) as file:
        lines = file.readlines()
        lines = [line.rstrip() for line in lines]
    return lines

In [10]:
input = ReadFileIntoList("input.txt")

In [11]:
import numpy as np
import pandas as pd

In [12]:
drawNumbers = [int(num) for num in input[0].split(',')]

In [13]:
boardsCollated = input[2:]
boards = [[]]
boardNumber = 0
for i in range(len(boardsCollated)):
    if boardsCollated[i] != '':
        boards[boardNumber].append([int(line) for line in boardsCollated[i].split(' ') if line != ''])
    if boardsCollated[i] == '':
        boards.append([])
        boardNumber += 1

In [14]:
def RecreateArrays():
    dfArrayToChange = []
    dfArrayToKeep = []
    for i in range(len(boards)):
        df = pd.DataFrame(boards[i])
        dfArrayToChange.append(df.copy())
        dfArrayToKeep.append(df.copy())
    return (dfArrayToChange,dfArrayToKeep)

In [15]:
def CheckDfForWinners(df):
    for i in df.index:
        checkArray = df.iloc[i].values
        isRowWinner = all(flag == '' for flag in checkArray) 
        checkArray = df.iloc[:,i].values
        isColumnWinner = all(flag == '' for flag in checkArray)
        if isRowWinner or isColumnWinner:
            return True
        return False

In [17]:
(dfArrayToChange,dfArrayToKeep) = RecreateArrays()
hasWinner = False
for numberToTickOff in drawNumbers:
    for j in range(len(dfArrayToChange)):
        df = dfArrayToChange[j]
        df.replace(to_replace=numberToTickOff,value='',inplace=True)
        hasWinner = CheckDfForWinners(df)
        #print(hasWinner)
        if hasWinner == True:
            print("Winner found")
            print(df)
            print(f"This array number {j}")
            print(f"This was the last number drawn: {numberToTickOff}")
            break
    if hasWinner == True:
        break
sumOfUnchecked = np.sum(np.sum(dfArrayToChange[j].replace(to_replace='',value=0)))
answer = sumOfUnchecked * numberToTickOff
print(f"Answer is {sumOfUnchecked} * {numberToTickOff} = {answer}")

Winner found
    0   1   2   3   4
0                    
1  27  56  50  33  54
2  67      35  25  10
3  93          90   4
4  57          83    
This array number 93
This was the last number drawn: 92
Answer is 684 * 92 = 62928


In [151]:
(dfArrayToChange,dfArrayToKeep) = RecreateArrays()
hasWinner = False
knownWinners = []
lastKnownWinnerPosn = None

for i in range(len(drawNumbers)):
    numberToTickOff = drawNumbers[i]
    for j in range(len(dfArrayToChange)):
        if any(flag == j for flag in knownWinners) == True:
            break
        df = dfArrayToChange[j]
        df.replace(to_replace=numberToTickOff,value='',inplace=True)
        hasWinner = CheckDfForWinners(df)
        if hasWinner == True:
            lastKnownWinner = df
            lastKnownWinnerPosn = j
            print(f"df {j} is a winner")
    if lastKnownWinnerPosn != None and any(flag == lastKnownWinnerPosn for flag in knownWinners) == False:
        knownWinners.append(lastKnownWinnerPosn)

sumOfUnchecked = np.sum(np.sum(dfArrayToChange[lastKnownWinnerPosn].replace(to_replace='',value=0)))
answer = sumOfUnchecked * numberToTickOff
print(f"Answer is {sumOfUnchecked} * {numberToTickOff} = {answer}")
        

df 93 is a winner
df 30 is a winner
df 28 is a winner
df 6 is a winner
df 3 is a winner
df 0 is a winner
Answer is 403 * 54 = 21762


In [152]:
i


99

In [108]:
for df in dfArrayToChange:
    if CheckDfForWinners(df) == True:
        print(df)

    0   1   2   3   4
0                    
1  27  56  50  33  54
2  67      35  25  10
3  93          90   4
4  57          83    


In [78]:
dfArrayToChange[93].replace(to_replace='',value=0,inplace=True)

In [81]:
np.sum(np.sum(dfArrayToChange[93]))

684

In [7]:
import os, re

PATH = "input.txt"

def parse_input(fin):
    rexp = re.compile(r'\d+')
    fin = [list(map(int,rexp.findall(line))) for line in fin.read().split("\n\n")] # [[call number], [bingo sheet], []]

    return fin[0], fin[1:] 

def count_part1(call, sheets, mode):
    winner = [0 for i in range(len(sheets))] # list to track completed sheet
    sheets = [[[i, 0] for i in sheet] for sheet in sheets] # [[number, 0] * 25],[]]
    for i in call:
        for sheet_pos, sheet in enumerate(sheets):
            if winner[sheet_pos] == 0:
                for number_pos, number_drawn in enumerate(sheet):
                    if number_drawn[0] == i: # if called number is already drawn then change 0 to 1
                        number_drawn[1] = 1
                        column, row = divmod(number_pos, 5)
                        check_row = all([sheet[n+(5*column)][1] for n in range(5)])
                        check_collumn = all([sheet[row+(5*n)][1] for n in range(5)])
                        if check_row or check_collumn:
                            if mode == "first":
                                print("sheet pos: ", sheet_pos)
                                return sheet, i
                            else: # mode == "last"
                                if winner.count(0) == 1: # if this sheet is last one to win
                                    return sheet, i
                                else:
                                    winner[sheet_pos] = 1 # track completed sheet

with open(PATH, 'r') as file:
    call, sheets = parse_input(file)

sheet, just_called = count_part1(call, sheets, "first")
ans = sum([i[0] for i in sheet if i[1] == 0]) * just_called # add all uncalled number then multiply that by number just called
print('Part 1:', str(ans))
print("last_called: ", just_called)

#sheet, just_called = count_part1(call, sheets, "last")
#ans = sum([i[0] for i in sheet if i[1] == 0]) * just_called
#print('Part 2:', str(ans))

sheet pos:  48
Part 1: 31424
last_called:  32


In [3]:
sheet_pos

NameError: name 'sheet_pos' is not defined

Guessed 15732(too low),62928(too high)
Guessed 21762(too low)